proj_eclipse/src/main/java/smartmag/models/BoxModel.java

```diff
@@ -0,0 +1,245 @@
+package smartmag.models;
+
+import static ingsw_proj_magazzino.db.generated.Tables.BOX;
+import static ingsw_proj_magazzino.db.generated.Tables.PRODOTTO;
+
+import java.sql.SQLIntegrityConstraintViolationException;
+import java.util.HashMap;
+
+import ingsw_proj_magazzino.db.generated.tables.records.BoxRecord;
+import ingsw_proj_magazzino.db.generated.tables.records.ProdottoRecord;
+import smartmag.data.Box;
+import smartmag.data.Prodotto;
+
+public class BoxModel extends BaseModel {
+
+	private static HashMap<String, BoxModel> instances = new HashMap<String, BoxModel>();
+	private Box box;
+	private BoxRecord record;
+
+	/**
+	 * restituisce il modello del box passato come parametro e lo salva
+	 * all'interno dell'hashmap che contiene i modelli di tutti i box
+	 * 
+	 * @param b box del quale si vuole ottenere il modello
+	 * @return modello del box
+	 */
+	public static BoxModel getBoxModel(Box b) {
+		if (b != null && b.isValid()) {
+			if (!instances.containsKey(b.getIndirizzo())) {
+				if ((DSL.select().from(PRODOTTO)
+						.where(PRODOTTO.ID.eq(b.getProd().getId()))) != null) {
+					BoxModel bm = new BoxModel(b);
+					instances.put(b.getIndirizzo(), bm);
+					return bm;
+				} else {
+					throw new IllegalArgumentException(
+							"prodotto non esistente");
+				}
+
+			} else {
+				return instances.get(b.getIndirizzo());
+			}
+		} else
+			throw new IllegalArgumentException("box non valido");
+	}
+
+	/**
+	 * metodo costruttore della classe BoxModel
+	 * 
+	 * @param b box del quale si vuole generare il modello
+	 */
+	private BoxModel(Box b) {
+		this.record = fetchBoxByIndirizzo(b.getIndirizzo());
+		this.box = b;
+	}
+
+	/**
+	 * controlla che ci sia un record del box nel db
+	 * 
+	 * @return true se é presente, false se non lo é
+	 */
+	public boolean isSavedInDb() {
+		refreshFromDb();
+		return record != null;
+	}
+
+	/**
+	 * restituisce l'oggetto Box rappresentato dal modello
+	 * 
+	 * @return
+	 */
+	protected Box getBox() {
+		return box;
+	}
+
+	protected void setBox(Box b) {
+		if (b != null && b.isValid() && (DSL.select().from(PRODOTTO)
+				.where(PRODOTTO.ID.eq(b.getProd().getId())) != null)) {
+			this.box = b;
+		} else
+			throw new IllegalArgumentException("box o prodotto non valido");
+
+	}
+
+	public BoxRecord getRecord() {
```

Comment:
Il record dovrebbe essere gestito esclusivamente dal modello stesso; al più lo metterei _protected_ se per caso dovesse servire in altri modelli mancanti tipo _Movimentazione_.